triton/tut/parallel-shared.rst

```diff
@@ -0,0 +1,257 @@
+===========================================================
+Shared memory parallelism: multithreading & multiprocessing
+===========================================================
+
+
+.. admonition:: Abstract
+
+   * Verify that your program can utilize multiple CPUs.
+
+   * Use ``--cpus-per-task=C`` to reserve multiple CPUs for your job.
+
+   * If you use ``srun`` to launch your program in your sbatch-script and
+     want your program to utilize all of the allocated CPUs, run
+     ``export SRUN_CPUS_PER_TASK=$SLURM_CPUS_PER_TASK`` in your script
+     before calling ``srun``.
+
+   * You must always :doc:`monitor jobs <monitoring>` to make sure they are using all the
+     resources you request (``seff JOBID``).
+
+   * If you aren't fully sure of how to scale up, contact us
+     :doc:`Research Software Engineers </rse/index>` early.
+
+What is shared memory parallelism?
+----------------------------------
+
+In shared memory parallelism a program will launch multiple processes
+or threads so that it can leverage multiple CPUs available in the machine.
+
+**Slurm reservations for both methods behave similarly.** This document
+will talk about processes, but everything mentioned would be applicable to threads
+as well. See section on
+:ref:`difference between processes and threads<threads-vs-processes>`
+for more information on who proceseses and threads differ.
+
+Communication between processes happends via shared memory. This means
+that all processes need to run on the same machine.
+
+.. figure:: /images/parallel-shared.png
+    :width: 80%
+    :align: center
+
+
+Depending on a program, you might have multiple processes (Matlab parallel
+pool, R parallel-library, Python multiprocessing) or have multiple threads
+(OpenMP threads of BLAS libraries that R/numpy use).
+
+Running a typical multiprocess program
+--------------------------------------
+
+Reserving resources for shared memory programs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Reserving resources for shared memory programs is easy: you'll only need to
+specify how many CPUs you want via ``--cpus-per-task=C``-flag.
+
+For most programs using ``--mem=M`` is the correct way to reserve memory,
+but in some cases the amount of memory needed scales with the number of
+processors. This might happen, for example, if each process opens a different
+dataset or runs its own simulation that needs extra memory allocations.
+In these cases you can use ``--mem-per-core=M`` to specify a memory allocation
+that scales with the number of CPUs. We recommend starting with ``--mem=M``
+if you do not know how your program scales.
+
+
+Running an example shared memory parallel program
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For this example, let's consider
+`pi.py <https://github.com/AaltoSciComp/hpc-examples/blob/master/slurm/pi.py>`_-example
+from the
+`hpc-examples <https://github.com/AaltoSciComp/hpc-examples>`_-repository.
```

Comment:
```suggestion
`pi.py <https://github.com/AaltoSciComp/hpc-examples/blob/master/slurm/pi.py>`__-example
from the
`hpc-examples <https://github.com/AaltoSciComp/hpc-examples>`__-repository.
```
Note: raw external links have two underscores after them.  One actually works, and I think it does something like define the text (`hpc-examples`) as an alias for the link itself, which can actually somehow be used in other links without re-definining.  I've never used this, though...